applications/ParticleMechanicsApplication/custom_io/particle_vtk_output.cpp

```diff
@@ -0,0 +1,430 @@
+            WriteModelPartToFile(r_sub_model_part, true, rOutputFilename);
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void ParticleVtkOutput::WriteModelPartToFile(
+    const ModelPart& rModelPart,
+    const bool IsSubModelPart,
+    const std::string& rOutputFilename
+    )
+{
+    // Make the file stream object
+    const std::string output_file_name = GetOutputFileName(rModelPart, IsSubModelPart, rOutputFilename);
+    std::ofstream output_file;
+    if (mFileFormat == ParticleVtkOutput::FileFormat::VTK_ASCII) {
+        output_file.open(output_file_name, std::ios::out | std::ios::trunc);
+        output_file << std::scientific;
+        output_file << std::setprecision(mDefaultPrecision);
+    } else if (mFileFormat == ParticleVtkOutput::FileFormat::VTK_BINARY) {
+        output_file.open(output_file_name, std::ios::out | std::ios::binary | std::ios::trunc);
+    }
+
+    KRATOS_ERROR_IF_NOT(output_file.is_open()) << "File \"" << output_file_name << "\" could not be opened!" << std::endl;
+
+    WriteHeaderToFile(rModelPart, output_file);
+    WriteMeshToFile(rModelPart, output_file);
+    if (mOutputEntities == ParticleVtkOutput::OutputEntities::ELEMENTS) {
+        WriteElementResultsToFile(rModelPart, output_file);
+    } else if (mOutputEntities == ParticleVtkOutput::OutputEntities::CONDITIONS) {
+        WriteConditionResultsToFile(rModelPart, output_file);
+    }
+
+    output_file.close();
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void ParticleVtkOutput::WriteMeshToFile(
+    const ModelPart& rModelPart,
+    std::ofstream& rFileStream
+    ) const
+{
+    WriteConditionsAndElementsToFile(rModelPart, rFileStream);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void ParticleVtkOutput::WriteConditionsAndElementsToFile(
+    const ModelPart& rModelPart,
+    std::ofstream& rFileStream
+    ) const
+{
+    const auto& r_local_mesh = rModelPart.GetCommunicator().LocalMesh();
+
+    if (mOutputEntities == ParticleVtkOutput::OutputEntities::ELEMENTS) {
+        const SizeType num_elements = rModelPart.GetCommunicator().GetDataCommunicator().SumAll(static_cast<int>(r_local_mesh.NumberOfElements()));
+        // Write points data
+        rFileStream << "POINTS " << num_elements << " float\n";
+        for (auto itr_element = r_local_mesh.ElementsBegin(); itr_element != r_local_mesh.ElementsEnd(); itr_element++) {
+            std::vector<array_1d<double, 3>> mp_coord = { ZeroVector(3) };
+            itr_element->CalculateOnIntegrationPoints(MP_COORD, mp_coord, rModelPart.GetProcessInfo());
```

Comment:
at first glance this seems to be the only change, the rest looks very similar to the Core Output

If there was a function `WritePoints`, could the rest of the code then be removed?

I would advise to avoid the duplication of code, as some of the writing things are quite delicate

Reply:
I removed part of the duplicated code. I cannot remove methods `WritePropertiesIdsToFile` and `WriteIdsToFile` because they are defined as private in the `VtkOutput` class.

What do you mean when you say that a function `WritePoints` could be sufficient? Should I write everything (`POINTS`, `CELLS`, `CELL_TYPES` and `POINT_DATA`) in this function?

Reply:
Let me reformulate my question, what is different for tze mpm vtk output?

I.e. what is the minimal changes necessary go make it work?

Reply:
The function `WriteNodesToFile` must be changed because the nodes now are the coordinates of the material points (this basically is the line of code you highlighted above).

Moreover, I have changed the function `WriteConditionsAndElementsToFile` because in the `VtkOutput` implemented in the core if a model part has both elements and conditions then only the elements are written, while in the implementation I proposed here the entity that is written (elements or conditions) is the one specified in the input parameters. As far as I have understood, it is not possible to write conditions when both elements and conditions are defined in a model part and this should be why `particle_vtk_condition_output_process.py` was implemented before. For the same reason it is also required to change the function `WriteConditionResultsToFile`.

Concerning the functions `WritePropertiesIdsToFile` and `WriteIdsToFile`, as I said in the previous comment, they are copied only because in the base class they are defined as private.

Reply:
Ok understood

Have a look at #11156, I think with this all you need to do is to override the `WriteNodesToFile` method right?

Reply:
I think so, but now there is a new problem.

I have tried to merge your branch `core/vtk-enhancements` to this one locally and to override only the method `WriteNodesToFile`. Kratos is compiling and the process is working without returning errors but data is not written in the VTK files.

I guess this is because the mpm model part that is used to initialize the `ParticleVtkOutput` has no elements and conditions in that moment (they are created in the `Initialize` stage of the `PythonSolver`, so after all the processes have already been defined). Then, even if in the parameters we specify `"entity_type" : "element"`, the value is of `mEntityType` is set to `EntityType::NONE` because the number of elements is zero:

https://github.com/KratosMultiphysics/Kratos/blob/63814884db4b81537ea46c814cec736c2a0a1d4d/kratos/input_output/vtk_output.cpp#L142-L144

I do not know if it is better to adjust your implementation in `core/vtk-enhancements` or if it is better to generate material point elements and conditions in an earlier stage of the simulation.

Reply:
Ok I see

I did not consider this case, I will update the implementation accordingly

Reply:
This would be a nice addition for me too... :D... I was also thinking to do the same... :D

Reply:
now it should work :)

Reply:
Ok, the class is writing data on the VTK files but there are other problems 😅

I think they are both related to the fact that the geometry type of mpm elements and conditions is `Kratos_Quadrature_Point_Geometry`.

When the method `WriteConnectivity` is called, the nodes of the background geometry are written instead of just a number referring to the points in the `POINTS` section

Moreover, when the method `WriteCellType` is called, the following error is shown

https://github.com/KratosMultiphysics/Kratos/blob/0affec97cbeabff2f75f07fddda0b66ff1767ed8/kratos/input_output/vtk_output.cpp#L425-L427

because `Kratos_Quadrature_Point_Geometry` is not included here

https://github.com/KratosMultiphysics/Kratos/blob/0affec97cbeabff2f75f07fddda0b66ff1767ed8/kratos/input_output/vtk_output.cpp#L64-L66

Can I also override this two methods?

Reply:
Do you really need to override? Would adding this type to the map be enough?

Reply:
For `WriteCellType` yes, it should be enough to add `Kratos_Quadrature_Point_Geometry` to `geo_type_vtk_cell_type_map` (didn't know if for you it's ok to add this in the core).

But this does not solve the problem related to `WriteConnectivity`.

Reply:
looking at the code I think you need to implement your own `WriteConnectivity` (the one in the baseclass is templated and can thus not be overridden)

I guess you can do this in this function (`WriteConditionsAndElementsToFile`), or have a `WriteParticleConnectivity`, which is called from here

Reply:
Summarizing, now I'm using the following methods:
* `WriteNodesToFile`: it writes the coordinates of the material points elements/conditions (virtual method in the core since #11156)
* `WriteCellType`: it is necessary since `Kratos_Quadrature_Point_Geometry` is not contained in `vtk_definitions.cpp`. I can add it to the map `VtkDefinitions::KratosVtkGeometryTypes` and remove this method but in my opinion in the map there should be only geometries that are compatible with the vtk class in the core, while `Kratos_Quadrature_Point_Geometry` requires other methods for writing the coordinates and the connectivity
* `WriteConnectivity`: it is necessary since the base method writes the IDs of the nodes of the backgroud mesh instead of the ones of the elements
* `WriteConditionsAndElementsToFile`: used to call the new methods `WriteConnectivity` and `WriteCellType` (requires to be defined as `virtual` in the core with a separate PR)
* `WriteNodalResultsToFile`: empty method that overrides the one in the core (requires to be defined as `virtual` in the core with a separate PR). This is necessary because, when the method in the base class is called, even if we don't want to write nodal results (like in mpm) it still writes `POINT_DATA number_of_nodes` and an error occur when trying to open the file with paraview

Reply:
sounds good to me

Please feel free to make the methods `virtual` if you need them to be, and tag me in the PR

Reply:
Done in #11482

Reply:
I won't suggest rewriting the process, but I think the logic of this output deviates from `VtkOutput` so much that it doesn't make sense to derive from it. Default params are restricted instead of expanded, some functions (such as `WriteNodalResultsToFile`) are overridden to do nothing, etc. Furthermore, there isn't anything to be gained from dynamic polymorphism here. It would've probably been better to compose `VtkOutput` instead of inheriting from it.

Reply:
Sorry, but I am not sure to have understood how you wold like the process to be implemented. 
Could you please provide me an example of something similar implemented in kratos?

Reply:
It's just a suggestion for the future, you don't need to change this now.

```diff
- class ParticleVtkOutput : public VtkOutput
+ class ParticleVtkOutput
{
    /*...*/
+ private: 
+ VtkOutput mVtkOutput;
};
```

Reply:
Ok, thanks for your comment and suggestion, I will consider it in the future!

Reply:
@matekelemen makes full sense to me, back then the design with derivation seemed the more useful, but perhaps it is not the case here

Regardless, I would not focus too much on vtk, but rather on vtu in the future